DataStructures/BinaryHeap.cs

```diff
@@ -0,0 +1,251 @@
+        /// the backing list, removing the last element, and pushing the new root down
+        /// until the heap property is restored.
+        /// </remarks>
+        /// <returns>The top/root of the heap.</returns>
+        /// <exception cref="InvalidOperationException">Thrown if heap is empty.</exception>
+        public T Peek()
+        {
+            if (Count == 0)
+            {
+                throw new InvalidOperationException("Heap is empty!");
+            }
+
+            return data[0];
+        }
+
+        /// <summary>
+        /// Returns element if it compares larger to the top/root of the heap, else
+        /// inserts element into the heap and returns the top/root of the heap.
+        /// </summary>
+        /// <param name="element">The element to check/insert.</param>
+        /// <returns>element if element compares larger than top/root of heap, top/root of heap otherwise.</returns>
+        public T PushPop(T element)
+        {
+            if (Count == 0)
+            {
+                return element;
+            }
+
+            if (comparer.Compare(element, data[0]) < 0)
+            {
+                T tmp = data[0];
+                data[0] = element;
+                HeapifyDown(0);
+                return tmp;
+            }
+            else
+            {
+                return element;
+            }
+        }
+
+        /// <summary>
+        /// Check if element is in the heap.
+        /// </summary>
+        /// <param name="element">The element to check for.</param>
+        /// <returns>true if element is in the heap, false otherwise.</returns>
+        public bool Contains(T element) => data.Contains(element);
+
+        /// <summary>
+        /// Remove an element from the heap.
+        /// </summary>
+        /// <remarks>
+        /// In removing an element from anywhere in the heap, we only need to push down or up
+        /// the replacement value depending on how the removed value compares to its
+        /// replacment value.
+        /// </remarks>
+        /// <param name="element">The element to remove from the heap.</param>
+        /// <exception cref="ArgumentException">Thrown if element is not in heap.</exception>
+        public void Remove(T element)
+        {
+            var idx = data.IndexOf(element);
+
+            if (idx == -1)
+            {
+                throw new ArgumentException($"{element} not in heap!");
+            }
+            else
+            {
+                Swap(idx, data.Count - 1);
+                T tmp = data[data.Count - 1];
+                data.RemoveAt(data.Count - 1);
+
+                if (idx < data.Count)
+                {
+                    if (comparer.Compare(tmp, data[idx]) > 0)
+                    {
+                        HeapifyDown(idx);
+                    }
+                    else
+                    {
+                        HeapifyUp(idx);
+                    }
+                }
+            }
```

Comment:
```suggestion
                Swap(idx, data.Count - 1);
                T tmp = data[data.Count - 1];
                data.RemoveAt(data.Count - 1);

                if (idx < data.Count)
                {
                    if (comparer.Compare(tmp, data[idx]) > 0)
                    {
                        HeapifyDown(idx);
                    }
                    else
                    {
                        HeapifyUp(idx);
                    }
                }
```
Else clause not needed, code won't execute any further after throw